### ISSUE TITLE
cuthalion0x - Liquidated collateral can get stuck in credit accounts

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-05-ironbank-judging#304]
cuthalion0x

high

# Liquidated collateral can get stuck in credit accounts

## Summary

When liquidating collateral via `IronBank#liquidate()`, the `liquidator` receives the liquidated `IBToken` amount. It is possible for the `liquidator` to be a credit account, and credit accounts are not properly equipped to handle `IBToken` transfers or redemption. So, the received `IBToken` amount may be stuck until such time as the credit account is able to pay down its debt.

## Vulnerability Detail

The `IronBank` designates "credit accounts" which exist outside of the typical supply/borrow/repay/redeem workflow. These accounts are uncollateralized, meaning they can borrow up to a given credit limit without supplying any collateral, and they are also immune to liquidation. They are trusted to repay their own debts over time.

Given the special treatment of these accounts, some important specializations are made across the `IronBank`:
- A credit account cannot be _supplied_ to, meaning it cannot mint `IBToken` collateral.
- A credit account does not have its `IBToken` collateral checked upon _borrowing_ because it does not need to collateralize loans.
- A credit account can only _borrow to_ and _repay for_ itself.
- A credit account cannot be _liquidated_.
- A credit account cannot _receive_ `IBToken` transfers executed via ERC20 methods (`IBToken#transfer()` and `IBToken#transferFrom()`).

However, there is no special treatment given to the following:
- A credit account attempting a _redemption_ will have its liquidity checked just like any other account (`collateral >= debt`).
- A credit account attempting to _send_ `IBToken` collateral will have its liquidity checked just like any other account.

There are no specializations made for sending or redeeming collateral tokens because credit accounts are not expected ever to hold `IBToken` collateral. They cannot mint it, and they cannot receive it via traditional transfers. However, there is one corner case in which they can receive it: **liquidations**.

When calling `IronBank#liquidate()`, the `borrower` is prohibited from being a credit account (because they cannot be liquidated), but the `liquidator` can be a credit account. This means that a credit account can receive `IBToken` collateral via liquidation even though it is never supposed to hold these tokens. This occurs via the call to `IronBank#_transferIBToken()`, which lacks the `isCreditAccount()` check from its external entrypoint, `IronBank#transferIBToken()`.

Once the credit account receives collateral, it may not be empowered to redeem that collateral. This is because:

1. Any attempt to call `IronBank#redeem()` will require an evaluation of the collateral/debt ratio via `_checkAccountLiquidity()`.
2. Any attempt to call `IBToken#transfer()` to move tokens to another address for redemption will also invoke `_checkAccountLiquidity()`.

This is problematic for credit accounts in particular because they are not supposed to maintain collateral for their borrow positions. As such, the credit account may very well have a large outstanding borrow, and it will never be able to redeem the liquidated collateral until it first pays its debt in full. The `IBToken` collateral is stuck for an unknown amount of time, which may be quite long depending on the conditions of the loan.

## Impact

`IBToken` collateral, received by the user as a reward for liquidating an at-risk account, can remain stuck indefinitely in a credit account. The user may never receive their reward for maintaining the system's integrity, or may receive it after an unknowably long time delay. This breaks the system's incentive model for liquidators.

## Code Snippet

Internal transfer logic (no check for credit accounts):

https://github.com/sherlock-audit/2023-05-ironbank/blob/main/ib-v2/src/protocol/pool/IronBank.sol#L865-L887

Liquidation (no check for credit account `liquidator`):

https://github.com/sherlock-audit/2023-05-ironbank/blob/main/ib-v2/src/protocol/pool/IronBank.sol#L480-L510

External transfer logic (checks account liquidity on L571):

https://github.com/sherlock-audit/2023-05-ironbank/blob/main/ib-v2/src/protocol/pool/IronBank.sol#L560-L572

Redemption logic (checks account liquidity on L448):

https://github.com/sherlock-audit/2023-05-ironbank/blob/main/ib-v2/src/protocol/pool/IronBank.sol#L406-L451

## Tool used

Manual Review

## Recommendation

Either add the following line to `IronBank#liquidate()`:

```solidity
        require(!isCreditAccount(liquidator), "cannot liquidate to a credit account");
```

Or consider moving the external transfer restriction on L566 of `IronBank#transferIBToken()` to the internal transfer function, `IronBank#_transferIBToken()`, to account for all transfers including liquidations:

https://github.com/sherlock-audit/2023-05-ironbank/blob/main/ib-v2/src/protocol/pool/IronBank.sol#L566

Duplicate of #458

**Comment by cuthalion0x** [comment]
Escalate for 10 USDC

First, this issue should be a duplicate of #458. Second, I disagree with the sponsor's conclusion mentioned in #458:

> The credit account can still redeem them to the underlying.

The credit account most likely **cannot** redeem the collateral without first paying down its debt, and that is the crux of the issue.

A credit account is designed to own significant debt with zero collateral. The redemption process does not account for the special case of credit accounts, and collateral can only be redeemed if an account is in good standing (health check on collateral/debt ratio). So, once this account receives a balance of `IBToken`, it very likely cannot redeem those tokens for the foreseeable future; the funds are effectively frozen. For example, the credit account may have intended to use these liquidated funds to repay some of its own debt, but it cannot actually do this because the debt must be paid _prior to redemption_.

This could be dismissed as user error on the liquidator's part, but to do so would be very negligent in my opinion as this outcome is not easily predictable. It is much safer - and quite easy - to restrict credit accounts from accidentally reaching this error state.

**Comment by 0xffff11** [comment]
I agree with duplication of #458 I appreciate the extra context on the sponsors comment on 458, and it does makes sense to me. I still think that is this statement is valid "The credit account can still redeem them to the underlying." it should be invalid, if not I see as a medium

**Comment by ibsunhub** [comment]
fix: https://github.com/ibdotxyz/ib-v2/pull/55

**Comment by hrishibhat** [comment]
Result:
Low
Duplicate of #458 
The Credit account is trusted and not expected to liquidate. 
More context from the protocol comments here:
https://github.com/sherlock-audit/2023-05-ironbank-judging/issues/98#issuecomment-1639594847

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [cuthalion0x](https://github.com/sherlock-audit/2023-05-ironbank-judging/issues/304/#issuecomment-1606092203): accepted

**Comment by IAm0x52** [comment]
Fix looks good. Same fix as #98
